File: source/bxgenbb_help/src/decay0_fortran/funbeta1.f
c funbeta1.f 
c
c This file was extracted from the 'decay0' program by V.I. Tretyak
c
c Copyright 1995-2011 V.I. Tretyak
c
c This program is free software; you can redistribute it and/or modify
c it under the terms of the GNU General Public License as published by
c the Free Software Foundation; either version 3 of the License, or (at
c your option) any later version.
c 
c This program is distributed in the hope that it will be useful, but
c WITHOUT ANY WARRANTY; without even the implied warranty of
c MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
c General Public License for more details.
c 
c You should have received a copy of the GNU General Public License
c along with this program; if not, write to the Free Software
c Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
c 
c
	function funbeta1(E)
c Function to search the maximum of beta spectrum curve by tgold
c subroutine.
c VIT, 30.07.1992; 15.10.1995; 31.03.2006.
	common/const/pi,emass,datamass(50)
	common/parbeta/Zdtr,Qbeta
	common/parbeta1/c1,c2,c3,c4
	funbeta1=0.
	if(E.gt.0.) then
c allowed spectrum
	   all=sqrt(E*(E+2.*emass))*(E+emass)*(Qbeta-E)**2*fermi(Zdtr,E)
c correction factor
	   w=E/emass+1.
	   cf=1.+c1/w+c2*w+c3*w**2+c4*w**3
c spectrum with correction
	   funbeta1=all*cf
	endif
	return
	end
c
c end of funbeta1.f
c
c Local Variables: --
c mode: fortran --
c End: --

File: source/bxgenbb_help/src/decay0/funbeta1.cc
// -*- mode: c++; -*- 
// funbeta1.cc

#include <iostream>
#include <cmath>
#include <stdexcept>

#include <gsl/gsl_math.h>

#include <genbb_help/decay0/funbeta1.h>
#include <genbb_help/decay0/fermi.h>
#include <genbb_help/decay0/particle.h>

namespace genbb {
  namespace decay0 {
     
    /// \brief Function to search the maximum of beta spectrum curve by decay0_tgold subroutine
    double decay0_funbeta1(double E_, void * params_)
    {
      const parbeta1 * pars = static_cast<const parbeta1 *>(params_);
      const double & Zdtr   = pars->Zdtr;
      const double & Qbeta  = pars->Qbeta;
      const double & c1     = pars->c1;
      const double & c2     = pars->c2;
      const double & c3     = pars->c3;
      const double & c4     = pars->c4;

      static const double emass = decay0_emass();
      double f = 0.0;
      if (E_ > 0.) {
        // allowed spectrum:
        double all = std::sqrt(E_ * (E_ + 2. * emass)) * (E_ + emass)
          *gsl_pow_2(Qbeta - E_) * decay0_fermi(Zdtr, E_);
        // correction factor:
        double w = E_ / emass + 1.;
        double cf = 1. 
          + c1 / w
          + c2 * w 
          + c3 * gsl_pow_2(w)
          + c4 * gsl_pow_3(w);
        // spectrum with correction:
        f = all * cf;
      }
      return f;
    }

  } // end of namespace decay0 
} // end of namespace genbb 
